### DISCUSSION TITLE
`fpm` on PyPi

### DISCUSSION THREAD

**Discussion by gnikit** [fortran-lang/fpm#696] [Ideas]
Have you considered releasing `fpm` through PyPi? _I think_ there is a way to get the conda-forge feedstock to also publish on PyPi.

While trying to see if `fpm` is available through PyPi I also saw that someone is name-squatting on the name, along with a plethora of other very common names, [see](https://pypi.org/user/eaybek/). Regardless of if you publish on PyPi it might be a good idea to file a [PEP 451 request](https://github.com/pypa/pypi-support/issues/new/choose) to reserve `fpm` for the organisation.

For more see: https://github.com/pypa/warehouse/issues/4004

**Comment by awvwgk** [comment]
Why would we want to `pip install fpm`? I know that CMake and ninja are doing this by building wheels with additional Python wrappers which than invoke the real binary. I have been using those in CI a bit, but eventually felt that it was easier to just install it directly via `conda`.

**Reply by gnikit** [reply]
My thinking was that not everyone uses Anaconda and Spack is not that common amongst individuals (afaik), so if you want an easy to use installer that is _relatively_ platform-agnostic pip is a good choice. Also, having to activate an Anaconda environment everytime you need to build from source sounds a little too involved for people who are used to just calling `make` or `cmake .. && make`.

Anyway, that's my 2 cents (and yes, I was thinking of CMake when I wrote this). IMO, it's easier to get people to use fpm if there barrier to entry is simply running `pip install fpm` and never having to think about it again, instead of having to have Anaconda, Spack, or directly downloading from GitHub.

**Reply by awvwgk** [reply]
Note that `pip install cmake` can fail, at least for me it did several times, especially on Windows. This is certainly not a surefire installation method, but we can try to create an `fpm` wheel as part of the CI workflow in this repo or as a separate repo.

**Reply by gnikit** [reply]
Yeah I think that sounds like a good idea. Having the `pip` option is always nice IMO.

**Reply by awvwgk** [reply]
Feel free to submit a patch regarding wheel builds. If you need help to claim the name on PyPI, we can file a request.

**Reply by gnikit** [reply]
BTW @awvwgk I feel I should provide some context as to why we would "need" pip instead of Anaconda. It is trivial to install packages with pip in VS Code so that would be an easy way to install `fpm` to people's machines. Anaconda on the other hand would be a nightmare since we would have to make our vscode-fortran-support extension manage conda environments.

I will try and have a look by the end of the week. Cheers for the input!

**Reply by awvwgk** [reply]
Supporting [`micromamba`](https://github.com/mamba-org/mamba#micromamba) as VS Code extensions seems like an effort very much worth the time then. Fitting a binary package in a Python wheel is a very crude workaround IMO.

**Reply by gnikit** [reply]
The problem I see is the added complexity to the extension. We would effectively have to rewrite a subset of the VS Code Python extension, which does support Anaconda. And then we would have to make people download Anaconda (and set it up correctly so that we can detect it from NodeJS).

micromamba seems very promising but from my experience users tend to get **very** upset when you install things on their system that they haven't heard about or don't think they are necessary. (also I think it will suffer from the same challenges as anaconda)

A big challenge  I see on the horizon with anaconda and micromamba is the need to communicate the initialised environment variables from the core of the extension, to the terminal windows at the bottom of VS Code, to the side panel actions and to the VS Code build system. That is a lot of work for something that is not integral to the extension and it's more a nice-to-have feature.

IMO pip is a very low bar to clear since it comes with Python nowadays. Installing fpm through pip is not probably the best idea if you want to deploy to an HPC but for local development in a user level I don't see any caveats. Are there any issues/conflicts that you can foresee with a `pip --user` installation?

PS I found this VS Code project that claims to allow for micromamba usage but it is in a prenatal state: https://marketplace.visualstudio.com/items?itemName=corker.vscode-micromamba

**Reply by awvwgk** [reply]
I have no experience with pip installing software outside of conda environments, I had to fix a few Python installations of coworkers who did break their machine with `[sudo] pip install` from time to time, but those are more anecdotal.

**Reply by gnikit** [reply]
Yeah, I wouldn't expect us doing anything as destructive as this. In VS Code specifically we run `{default-python} -m pip install --user --upgrade {package-name}` which installs locally and does not break anything. Also because `fpm` is not a Python package and has virtually no Python dependencies even if we wanted we couldn't break peoples' packages.


I was hoping we could be lazy and just piggyback on the existing conda-forge binaries but from the looks of CMake [1] we will have to be way more explicit.

Do you see another way of doing this? AFAIK as I know there is no easy way to go from Anaconda to PyPi.


[1]: https://github.com/scikit-build/cmake-python-distributions/blob/master/.github/workflows/build.yml
